2016-17/provincial/are/pipeline-spec.yaml

```diff
@@ -0,0 +1,183 @@
+are-2016-17:
+  title: Annual Report Expenditure 2016-17
+  description: Annual Report Expenditure, with Audited Outcome and Final Appropriation.
+  pipeline:
+  - run: load
+    parameters:
+      from: "https://s3-eu-west-1.amazonaws.com/manual-uploads.vulekamali.gov.za/are/PRV+AR+2016-17.xlsx"
+      name: 'are-2016-17'
+      format: 'xlsx'
+    cache: true
+
+  - run: delete_fields
+    parameters:
+      resources: are-2016-17
+      fields:
+      - BudYear
+      - ValueIndi
+      - Combo
+      - GFS1
+      - GFS2
+      - GFS3
+
+  - run: filter
+    parameters:
+      resources: are-2016-17
+      out:
+        - FY_Descript: Variance
+  - run: filter
+    parameters:
+      resources: are-2016-17
+      out:
+        - FY_Descript: Revised payments
+
+  - run: add_computed_field
+    parameters:
+      resources: are-2016-17
+      fields:
+      - operation: format
+        target: Budget Phase
+        with: '{FY_Descript}'
+
+  - run: find_replace
+    parameters:
+      resources: are-2016-17
+      fields:
+      - name: FinYear
+        patterns:
+        - find: (\d{4})_\d{2}
+          replace: \1
+      - name: Budget Phase
+        patterns:
+        - find: ^(Actual Payments)
+          replace: "Audit Outcome"
+        - find: ^Virement and shifting of funds
+          replace: "Final Appropriation"
+      - name: FY_Descript
+        patterns:
+        - find: ^(Actual Payments)
+          replace: Total
+        - find: ^(Adjusted Appropriation)
+          replace: Total
+        - find: ^(Final Appropriation)
+          replace: Total
+      - name: Department
+        patterns:
```

Comment:
mmmm not sure why this was in previous examples - please remove from this one

Reply:
See above

Reply:
Do you want me to remove all of the replacements for FY_Descript or only for Final Appropriation?

Reply:
No, please keep the replacements of FY_Descript. Only remove replacements of Department

Reply:
But when I do that, I face KeyError exception. Here's an example:

> INFO    :RESULTS:
> INFO    :FAILURE: ./2016-17/provincial/are/are-2016-17 
> ERROR log from processor programme_names:
> +--------
> | ERROR   :Traceback (most recent call last):
> | ERROR   :File "/home/baris/PycharmProjects/treasury-pipelines/programme_names.py", line 87, in <module>
> | ERROR   :process_row=process_row)
> | ERROR   :File "/home/baris/miniconda3/lib/python3.7/site-packages/datapackage_pipelines/wrapper/wrapper.py", line 232, in process
> | ERROR   :process_row)
> | ERROR   :File "/home/baris/miniconda3/lib/python3.7/site-packages/datapackage_pipelines/wrapper/wrapper.py", line 181, in __exit__
> | ERROR   :spew(self.datapackage, self.resource_iterator, stats=self.stats)
> | ERROR   :File "/home/baris/miniconda3/lib/python3.7/site-packages/datapackage_pipelines/wrapper/wrapper.py", line 75, in spew
> | ERROR   :for rec in res:
> | ERROR   :File "/home/baris/miniconda3/lib/python3.7/site-packages/datapackage_pipelines/wrapper/wrapper.py", line 201, in generic_process_resource
> | ERROR   :parameters, stats)
> | ERROR   :File "/home/baris/PycharmProjects/treasury-pipelines/programme_names.py", line 72, in process_row
> | ERROR   := programme_names[sphere][government_name][department_name].\
> | ERROR   :KeyError
> | ERROR   ::
> | ERROR   :'SOCIAL DEVELOPMENT AND SPECIAL PROGRAMMES'
> +--------

Reply:
Ah I see - please remove the programme_names processor then.

The reasoning here is, the data we get is dirty, but we don't have the authority to change department names. If it's just a matter of capitalisation we can and do "fix" it with "slugify". But even a spelling mistake we're not fixing at the moment. Instead, when there are discrepancies between department or programme names when comparisons are done later, our government can see the issue and fix their source data.

Thanks for trying though.

Reply:
Okay, I understand. You can find the changed versions of two files. When whole process is done, I'm going to send you the results as well.

https://github.com/vulekamali/treasury-pipelines/pull/25/commits/9eca39925701ff85e34c1e81d98675d8914119a4